### DISCUSSION TITLE
mogrify photoshop PSD to PNG throwing errors

### DISCUSSION THREAD

**Discussion by 5Diraptor** [ImageMagick/ImageMagick#2368] [Help] [answered]
OK I'm a noob and I only heard of ImageMagick today but decided to try it out.  I've written this batch command: 

 ```
%~d1
  CD "%~p1"
  MD "Studio High Res JPG"
  MD "Studio High Res PNG"
  MD "Studio Low Res JPG"
  MD "Studio Low Res PNG"
  MD "Studio Web JPG"
  MD "Studio Web PNG"
  FOR %%a in (*.psd) DO (
    magick mogrify -format jpg -path "Studio High Res JPG" -background white -alpha remove -alpha off -resize 4600x4600 *[0]
    magick mogrify -format png -path "Studio High Res PNG" -resize 1600x1600 *[0]
    magick mogrify -format jpg -path "Studio Low Res JPG" -background white -alpha remove -alpha off -resize 1600x1600 *[0]
    magick mogrify -format png -path "Studio Low Res PNG" -resize 1600x1600 *[0]
    magick mogrify -format jpg -path "Studio Web JPG" -background white -alpha remove -alpha off -resize 500x500 -gravity Center -extent 500x500 *[0]
    magick mogrify -format png -path "Studio Web PNG" -resize 500x500 -gravity Center -extent 500x500 *[0]
  )
  PAUSE
```
which seemed to work fine when I was testing on a couple of PSD files, but now I'm running it on a batch of 12 photoshop images and I'm getting errors - see screenshot below.  Am I missing something?
![image](https://user-images.githubusercontent.com/22928943/89297794-5d5e0700-d65c-11ea-9a42-a81bf3ff14f6.png)

**Comment by 5Diraptor** [comment]
CMD> magick -version results in:
Version: ImageMagick 7.0.10-25 Q16 x64 2020-08-01 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192628806
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib

**Comment by 5Diraptor** [comment]
changed the asterisk at the end of the command to be *.psd instead and it worked.  I think it was trying to convert directories for some reason as well as the rest.

**Comment by snibgo** [comment]
Your code is massively inefficient.

It loops through all of the PSD files ("FOR %%a in (*.psd)"). Then for each of the PSD file, it calls mogrify to process all of the PSD files.

I suggest you either remove the "FOR" loop, or change your "magick mogrify" commands to simple "magick" commands so that each IM command reads just one input, "%%a".

**Reply by 5Diraptor** [reply]
@snibgo Thanks!  I literally don't know what I'm doing and I'm making it up from templates I've researched so couple of quick questions if you don't mind:

- My first batch script I wrote used `magick convert` but the problem with that is when it saved the files as a new file, the original file extension was included, such as "image_name.psd.jpg" - mogrify is renaming files correctly at the moment.
- But I know there is something wrong, because I ended up putting a `PAUSE` at the end of the loop before the bracket, otherwise it reloops through a lot of times, but I didn't know enough to know why.
- If I remove the FOR loop, how do I make the code this?

` %~d1
  CD "%~p1"
  MD "Studio High Res JPG"
  MD "Studio High Res PNG"
  MD "Studio Low Res JPG"
  MD "Studio Low Res PNG"
  MD "Studio Web JPG"
  MD "Studio Web PNG"
  FOR %%a in (*.psd) DO (
    magick mogrify -format jpg -path "Studio High Res JPG" -background white -alpha remove -alpha off -resize 4600X4600 *.psd[0]
    magick mogrify -format png -path "Studio High Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Low Res JPG" -background white -alpha remove -alpha off -resize 1600X1600 *.psd[0]
    magick mogrify -format png -path "Studio Low Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Web JPG" -background white -alpha remove -alpha off -resize 500X500 -gravity Center -extent 500X500 *.psd[0]
    magick mogrify -format png -path "Studio Web PNG" -resize 500X500 -gravity Center -background transparent -extent 500X500 *.psd[0]
    PAUSE
  )
  PAUSE`

So when I was reading the documentation yesterday, it said that the way to process multiple commands on a file was through a FOR loop.  How would be better to do this?

**Comment by 5Diraptor** [comment]
Ah, I've just looked through documentation again and updated to this:

` %~d1
  CD "%~p1"
  MD "Studio High Res JPG"
  MD "Studio High Res PNG"
  MD "Studio Low Res JPG"
  MD "Studio Low Res PNG"
  MD "Studio Web JPG"
  MD "Studio Web PNG"
    magick mogrify -format jpg -path "Studio High Res JPG" -background white -alpha remove -alpha off -resize 4600X4600 *.psd[0]
    magick mogrify -format png -path "Studio High Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Low Res JPG" -background white -alpha remove -alpha off -resize 1600X1600 *.psd[0]
    magick mogrify -format png -path "Studio Low Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Web JPG" -background white -alpha remove -alpha off -resize 500X500 -gravity Center -extent 500X500 *.psd[0]
    magick mogrify -format png -path "Studio Web PNG" -resize 500X500 -gravity Center -background transparent -extent 500X500 *.psd[0]
  PAUSE`

and it seems to work well.  Is that an ideal script or are there anything else I can do to streamline it?

**Comment by snibgo** [comment]
I can't read your code. It comes out as an unbroken block of text. Please format it as you did in the OP.

**Reply by 5Diraptor** [reply]
%~d1
  CD "%~p1"
  MD "Studio High Res JPG"
  MD "Studio High Res PNG"
  MD "Studio Low Res JPG"
  MD "Studio Low Res PNG"
  MD "Studio Web JPG"
  MD "Studio Web PNG"
    magick mogrify -format jpg -path "Studio High Res JPG" -background white -alpha remove -alpha off -resize 4600X4600 *.psd[0]
    magick mogrify -format png -path "Studio High Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Low Res JPG" -background white -alpha remove -alpha off -resize 1600X1600 *.psd[0]
    magick mogrify -format png -path "Studio Low Res PNG" -resize 1600X1600 *.psd[0]
    magick mogrify -format jpg -path "Studio Web JPG" -background white -alpha remove -alpha off -resize 500X500 -gravity Center -extent 500X500 *.psd[0]
    magick mogrify -format png -path "Studio Web PNG" -resize 500X500 -gravity Center -background transparent -extent 500X500 *.psd[0]
  PAUSE

**Comment by snibgo** [comment, marked as answer]
That's better. It doesn't have the massive inefficiency.

However, you read each PSD file six times. If you use a "for" loop, you can use plain "magick" (instead of "magick mogrify") to read each PSD once, and resize and write six times. Windows BAT syntax:
```
FOR %%A in (*.psd) DO (

  magick %%A ^
    ^( -clone 0 ^
       -background white -alpha remove -alpha off -resize 4600x4600 ^
       -write "Studio High Res JPG\%%~nA.jpg" ^
       +delete ^
    ^) ^
    ^( -clone 0 ^
       -resize 1600x1600 ^
       -write "Studio High Res PNG\%%~nA.png" ^
       +delete ^
    ^) ^
    ^( -clone 0 ^
       -background white -alpha remove -alpha off -resize 1600x1600 ^
       -write "Studio Low Res JPG\%%~nA.jpg" ^
       +delete ^
    ^) ^
    ^( -clone 0 ^
       -resize 1600x1600 ^
       -write "Studio Low Res PNG\%%~nA.png" ^
       +delete ^
    ^) ^
    ^( -clone 0 ^
       -background white -alpha remove -alpha off -resize 500x500 -gravity Center -extent 500x500 ^
       -write "Studio Web JPG\%%~nA.jpg" ^
       +delete ^
    ^) ^
    ^( -clone 0 ^
       -resize 500x500 -gravity Center -extent 500x500 ^
       -write "Studio Web PNG\%%~nA.png" ^
       +delete ^
    ^) ^
    NULL:
)
```
Untested, so I've probably made a mistake.

For "%~nA", type "help for" at the command prompt.

Incidentally, I suggest not having spaces in directories or filenames. They just create problems.

EDIT: Added "NULL:" at the end.
